### PR TITLE
[Dataset & Transform] Synthetic Datasets for Explainability and SIGNDiffusion Transform

### PR DESCRIPTION
**Pull request by mufeili**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/data/explain_syn.py [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+        edges = list(g.edges())
+        src, dst = map(list, zip(*edges))
+        n = nx.number_of_nodes(g)
+
+        # Nodes in the base tree graph belong to class 0
+        node_labels = [0] * n
+        # The motifs will be evenly attached to the nodes in the base graph.
+        spacing = math.floor(n / self.num_motifs)
+
+        cycle_size = 6
```
Comment (jermainewang) [inline python/dgl/data/explain_syn.py:427]: Consider exposing this parameter

File: python/dgl/data/explain_syn.py [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+                a = np.random.randint(1, 4)
+                b = np.random.randint(1, 4)
+                src.append(n + a)
+                dst.append(anchor + b)
+
+            n += cycle_size
+
+        g = graph((src, dst), num_nodes=n)
+
+        # Perturb the graph by adding non-self-loop random edges
```
Comment (jermainewang) [inline python/dgl/data/explain_syn.py:454]: Do we have a utility function to perturb a graph?

File: python/dgl/data/explain_syn.py [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+        src, dst = map(list, zip(*edges))
+        n = nx.number_of_nodes(g)
+
+        # Nodes in the base tree graph belong to class 0
+        node_labels = [0] * n
+        # The motifs will be evenly attached to the nodes in the base graph.
+        spacing = math.floor(n / self.num_motifs)
+
+        # Construct a three-by-three grid
+        motif_g = nx.grid_graph([3, 3])
```
Comment (jermainewang) [inline python/dgl/data/explain_syn.py:590]: Consider expose number of grid graph nodes to argument.

File: python/dgl/data/explain_syn.py [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+
+    @property
+    def num_classes(self):
+        return 2
+
+class BA2Motifs(DGLBuiltinDataset):
+    r"""BA-2motifs dataset from `Parameterized Explainer for Graph Neural Network
+    <https://arxiv.org/abs/2011.04573>`__
+
+    This is a synthetic dataset of 800 graphs for graph classification. It was generated by
```
Comment (jermainewang) [inline python/dgl/data/explain_syn.py:664]:
```suggestion
    This is a synthetic dataset for graph classification. It was generated by
```

File: python/dgl/data/explain_syn.py (hunk 3 above)
Comment (jermainewang) [reply python/dgl/data/explain_syn.py:590]: I also suggest to add some graph transform utility such as `add_ring_motif`, `add_house_motif`, etc.

File: python/dgl/transforms/module.py [hunk 5]
```diff
@@ -1492,3 +1494,166 @@ def __call__(self, g):
...
+        which commonly lies in :math:`[0.05, 0.2]`.
+
+    Example
+    -------
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import SIGNDiffusion
+
+    >>> transform = SIGNDiffusion(k=2, eweight_name='w')
```
Comment (jermainewang) [inline python/dgl/transforms/module.py:1549]: If I want to specify multiple diffusion operators and save all of them into the graph, is there a way to do so in the transform interface? I won't suggest extending `diffuse_op` to a list, which may further complicate the code logic. I'm thinking of allowing users to change the output feature name.

File: python/dgl/transforms/module.py [hunk 6]
```diff
@@ -1492,3 +1494,166 @@ def __call__(self, g):
...
+                g.update_all(fn.u_mul_e(self.feat_name, eweight_name, 'm'),
+                             fn.sum('m', self.feat_name))
+                feat_list.append(g.ndata[self.feat_name])
+        return feat_list
+
+    def ppr(self, g):
+        feat_list = []
+        with g.local_scope():
+            eweight_name = 'w' if self.eweight_name is None else self.eweight_name
+            transform = GCNNorm(eweight_name=eweight_name)
```
Comment (jermainewang) [inline python/dgl/transforms/module.py:1649]: Don't we have a PPR data transform as well?

File: tests/compute/test_transform.py [hunk 7]
```diff
@@ -2350,6 +2352,15 @@ def test_module_laplacian_pe(idtype):
...
         assert F.allclose(new_g.ndata['lappe'].abs(), tgt)
 
+@pytest.mark.parametrize('g', get_cases(['has_scalar_e_feature']))
+@pytest.mark.parametrize('op', ['raw', 'rw', 'gcn', 'ppr'])
+def test_module_sign(g, op):
+    g = g.to(F.ctx())
+    transform = dgl.SIGNDiffusion(k=2, feat_name='h', diffuse_op=op)
+    transform(g)
+    transform = dgl.SIGNDiffusion(k=2, feat_name='h', eweight_name='scalar_w', diffuse_op=op)
+    transform(g)
```
Comment (jermainewang) [inline tests/compute/test_transform.py:2362]: Need more thorough unit test. I also suggest testing value correctness using some simple graph and features (such as all one).

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/data/__init__.py [hunk 8]
```diff
@@ -31,6 +31,7 @@
 from .fakenews import FakeNewsDataset
 from .csv_dataset import CSVDataset
 from .adapter import AsNodePredDataset, AsLinkPredDataset
+from .explain_syn import BAShapes, BACommunity, TreeCycles, TreeGrids, BA2Motifs
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/data/__init__.py]: Done

File: python/dgl/data/dgl_dataset.py [hunk 9]
```diff
@@ -313,6 +313,7 @@ def __init__(self, name, url, raw_dir=None, hash_key=(),
     def download(self):
         r""" Automatically download data and extract it.
         """
-        zip_file_path = os.path.join(self.raw_dir, self.name + '.zip')
-        download(self.url, path=zip_file_path)
-        extract_archive(zip_file_path, self.raw_path)
+        if self.url is not None:
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/data/dgl_dataset.py:316]: `DGLBuiltinDataset` is not exposed to users while `DGLDataset` is and inherits `DGLBuiltinDataset`. The default value of `url` for `DGLBuiltinDataset` has been `None` and I've updated its doc.

File: python/dgl/data/explain_syn.py [hunk 10]
```diff
@@ -0,0 +1,757 @@
...
+    following steps in order.
+
+    - Construct a base Barabási–Albert (BA) graph.
+    - Construct a set of five-node house-structured network motifs.
+    - Attach the motifs to randomly selected nodes of the base graph.
+    - Perturb the graph by adding random edges.
+    - Nodes are assigned to 4 classes. Nodes of label 0 belong to the base BA graph. Nodes of
+      label 1, 2, 3 are separately at the middle, bottom, or top of houses.
+    - Generate constant feature for all nodes, which is 1.
+
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/data/explain_syn.py]: Yes. What kind of images are you considering?

File: python/dgl/data/explain_syn.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/data/explain_syn.py]: This is the hyperparameter for the motif. The question is then if you also want to expose the motif hyperparameter for the other datasets.

File: python/dgl/data/explain_syn.py [hunk 11]
```diff
@@ -0,0 +1,757 @@
...
+    - Construct a base Barabási–Albert (BA) graph.
+    - Construct a set of five-node house-structured network motifs.
+    - Attach the motifs to randomly selected nodes of the base graph.
+    - Perturb the graph by adding random edges.
+    - Nodes are assigned to 4 classes. Nodes of label 0 belong to the base BA graph. Nodes of
+      label 1, 2, 3 are separately at the middle, bottom, or top of houses.
+    - Generate normally distributed features.
+    - Repeat the above steps to generate another graph. Its nodes are assigned to class
+      4, 5, 6, 7. Its node features are generated with a distinct normal distribution.
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/data/explain_syn.py]: Done.

File: python/dgl/data/explain_syn.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/data/explain_syn.py]: Done.

File: python/dgl/data/explain_syn.py [hunk 12]
```diff
@@ -0,0 +1,757 @@
+"""Synthetic datasets for developing GNN explainability approaches."""
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/data/explain_syn.py]: Done.

File: python/dgl/data/explain_syn.py (hunk 2 above)
Comment (mufeili) [reply python/dgl/data/explain_syn.py]: One possible solution is to extend [`AddEdge`](https://docs.dgl.ai/generated/dgl.transforms.AddEdge.html#dgl.transforms.AddEdge) transform.

File: python/dgl/data/explain_syn.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/data/explain_syn.py]: Done for exposing an argument controlling the number of grid graph nodes

Comment (mufeili) [reply python/dgl/data/explain_syn.py]: I'm not sure about the benefit of having utility functions like `add_ring_motif` and `add_house_motif`. I've only seen the use of them in constructing these synthetic datasets for GNN explainability. Also it's not clear how to attach motifs to a graph in a general way beyond these datasets.

File: python/dgl/transforms/module.py (hunk 5 above)
Comment (mufeili) [reply python/dgl/transforms/module.py:1549]: Replaced `feat_name` with `in_feat_name` and `out_feat_name`.

File: python/dgl/transforms/module.py (hunk 6 above)
Comment (mufeili) [reply python/dgl/transforms/module.py:1649]: This is essentially `APPNP`. However one can only get the output with a particular `K` with `APPNP`, not for `k=1,2,...,K`.

File: tests/compute/test_transform.py (hunk 7 above)
Comment (mufeili) [reply tests/compute/test_transform.py:2420]: Updated

File: python/dgl/data/explain_syn.py (hunk 3 above)
Comment (jermainewang) [reply python/dgl/data/explain_syn.py]: I see. That makes sense.

**Review by jermainewang — APPROVED** [review]
LGTM

**Pull request merged**